### PR TITLE
Autocast support for scaled_dot_product_attention

### PR DESCRIPTION
**Pull request by mikekgfb**
Summary: Autocast support for scaled_dot_product_attention

Test Plan: sandcastle and guthub cicd

Differential Revision: D42085525

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

File: aten/src/ATen/autocast_mode.cpp [hunk 1]
```diff
@@ -389,6 +389,8 @@ TORCH_LIBRARY_IMPL(aten, Autocast, m) {
   KERNEL(gru_cell, lower_precision_fp)
   KERNEL(rnn_tanh_cell, lower_precision_fp)
   KERNEL(rnn_relu_cell, lower_precision_fp)
+  KERNEL(_scaled_dot_product_flash_attention, lower_precision_fp)
+  KERNEL(_scaled_dot_product_attention, lower_precision_fp)
```
Comment (ngimel) [inline aten/src/ATen/autocast_mode.cpp:393]: I think we need only `_scaled_dot_product_flash_attention`, also we need a dedicated test (because previously this failure wasn't caught).

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/autocast_mode.cpp (hunk 1 above)
Comment (mikekgfb) [reply aten/src/ATen/autocast_mode.cpp:393]: I updated the PR to include tests that test that flash is failing when run with FP32 and works with autocast to float16 and bfloat16

File: test/test_transformers.py [hunk 2]
```diff
@@ -1292,6 +1292,19 @@ def test_unaligned_tensors(self):
             self.assertRaises(RuntimeError, lambda: torch.nn.functional._scaled_dot_product_attention(
                 q, k, v, None, 0.0, False, False))
 
+    @unittest.skipIf(not TEST_CUDA or not SM80OrLater or TEST_WITH_ROCM, "CUDA unavailable")
+    def test_flash_fail_fp32t(self):
+        device = 'cuda'
+        dtype = torch.float
+        size = (2, 2, 8, 5)
```
Comment (drisspg) [inline test/test_transformers.py:1299]: Flash doesn't accept head dim size of 5 so I don't this is failing for dtype 32 which we do already test with. If you decorate with auto cast does this enable flash?

**Review by drisspg — CHANGES_REQUESTED** [review]

**Review by drisspg — APPROVED** [review]
👍
